pennylane/spin/spin_hamiltonian.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains functions to create different templates of spin Hamiltonians.
+"""
+
+from pennylane import X, Z, math
+
+from .lattice import _generate_lattice
+
+# pylint: disable=too-many-arguments
+
+
+def transverse_ising(
+    lattice, n_cells, coupling=None, h=1.0, boundary_condition=False, neighbour_order=1
+):
+    r"""Generates the transverse field Ising model on a lattice.
+
+    The Hamiltonian is represented as:
+
+    .. math::
+
+        \hat{H} =  -J \sum_{<i,j>} \sigma_i^{z} \sigma_j^{z} - h\sum{i} \sigma_{i}^{x}
+
+    where ``J`` is the coupling defined for the Hamiltonian, ``h`` is the strength of the transverse
+    magnetic field and ``i,j`` represent the indices for neighbouring spins.
+
+    Args:
+       lattice (str): Shape of the lattice. Input Values can be ``'Chain'``, ``'Square'``,
+                      ``'Rectangle'``, ``'Honeycomb'``, ``'Triangle'``, or ``'Kagome'``.
+       n_cells (list[int]): Number of cells in each direction of the grid.
+       coupling (List[float] or List[math.array[float]]): Coupling between spins, it can be a
+                      list of length equal to ``neighbour_order`` or a square matrix of size
+                      ``(num_spins,  num_spins)``. Default value is [1.0].
+       h (float): Value of external magnetic field. Default is 1.0.
```

Comment:
This cannot be an array of length num_spins?

Reply:
This is external magnetic field and is usually taken as constant, I have not seen anything in literature where all spins experience different magnetic field. Code-wise it can be changed.